### PR TITLE
Checks if 'previousCssValue' exists

### LINKED ISSUE #52: Providing string to `css` binding breaks
# Issue

```ts
css: computed(() =>
  shouldShowPositiveFeedback.value ? 'valid-feedback' : 'invalid-feedback',
),
```
```
TypeError: Cannot read properties of undefined (reading 'split')
```

Initially, this value doesn't exists, so is `undefined`.

```ts
if (typeof classes === 'string') {
      // first remove the previously set classes, since our binding has changed
      const previousCssValue = (target as any)[previousCssBindingKey];
      target.classList.remove(...previousCssValue.split(/\s/gi));
```

# Fix

`remove` should only be executed if `previousCssValue` is set.

### PR DESCRIPTION
**Pull request by gimenezpablonl**
Check if 'previousCssValue' value is set when a string is provided to the css binding.

Previously when setting a string to the css binding, the application would throw an error because it tried to use the split() method on 'previousCssValue' which at first is an undefined variable.

### CODE REVIEW

**Review by ThaNarie — CHANGES_REQUESTED** [review]
The code looks good, however:

* Could you change your commit message according to https://cbea.ms/git-commit/, and force push it in this same branch?
* This wasn't specifically mentioned in the ticket, but could you add a test story with the different ways the css binding can be used, so we can be sure we don't break this in the future? In `test-storybook/src/components/bindings`.

Thanks!

**Review by ThaNarie — CHANGES_REQUESTED** [review]
Would you be able to update the stories so the CSS binding is actually updated, let's say by conditionally adding 0, 1 or 2 classes (using both string and object methods)?

Currently it's static, so the "removing when changing" is not really tested.

To make the tests even more useful, you could add an an existing classname to the `box` element in the template, and see if it gets persisted.

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        bind(refs.box, { css: { 'bg-primary': checked1, 'text-success': checked2, 'fs-3': checked3 } }),
+        bind(refs.checkbox1, { checked: checked1 }),
+        bind(refs.checkbox2, { checked: checked2 }),
+        bind(refs.checkbox3, { checked: checked3 }),
+      ];
+    },
+  }),
+  template: () => html` <div data-component="object">
+  <p>Css Object</p>
+    <div data-ref="box" class="bg-primary">
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:35]: With this suggestion:
> To make the tests even more useful, you could add an an existing classname to the box element in the template, and see if it gets persisted.

I meant an existing classname that is not controlled by the binding (one that should always stay there), like a `class=""leave-this-hear bg-primary"`.

Having an initial class that _is_ controlled by the binding there is also a good thing, so I would do both :)

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/* eslint-disable max-lines */
+import { html } from '@muban/template';
+import type { Story } from '@muban/storybook/types-6-0';
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:3]: For new stories I'd like to use the CFSv3 (object) format, so:

```ts
import type { Story } from '@muban/storybook';

export const CssObject: Story = {
  render() {
    return {
      component: ...
      template: ...
    }
  }
}
```

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
+/* eslint-disable max-lines */
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:1]: This comment can be removed from this file

Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:1]: This repo doesn't have pre-commit hooks set up, but could you at lease configure your editor to run eslint/prettier on save, so this files is properly formatted? :)

Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:1]: To make this test even more useful, I think we should add "interaction testing" using the `play` function.

More info: https://storybook.js.org/docs/react/writing-tests/interaction-testing
Example: https://github.com/mubanjs/muban/blob/ede37994c1cd8716b27951873d0be91e36e1c1fc/test-storybook/src/components/core/provideInject/ProvideInject.stories.ts#L210-L254

With that, we can (using code), change the checkboxes, and check if the classes are actually correctly added :)

**Review by ThaNarie — CHANGES_REQUESTED** [review]
Some suggestion to improve the way components are written, and how to add testing to them :)

**Comment by gimenezpablonl** [comment]
Last commit has a storybook related to #24. Can't reproduce that issue.

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    />
+    .text-success
+  </p>
+  <p>
+    <input data-ref="checkbox" data-testid="checkbox-fs-3" type="checkbox" value="fs-3" /> .fs-3
+  </p>
+</div>`;
+
+const interactiveTest = async () => {
+  await expect(screen.getByTestId('box').classList.value).toBe('box bg-primary');
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:52]: If you use `jest-dom`, you can use specific matchers using less code :)
https://github.com/testing-library/jest-dom#tohaveclass

Just make sure to install and `import '@testing-library/jest-dom'` at the top of the file

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  defineComponent,
+  computed,
+  ref,
+  refCollection,
+  ComponentFactory,
+  ComponentRefItem,
+} from '@muban/muban';
+import { screen, userEvent } from '@storybook/testing-library';
+import { expect } from '@storybook/jest';
+import type { Binding } from '../../../../../types/lib/bindings/bindings.types';
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:14]:
```suggestion
import type { Binding } from '@muban/muban/types/lib/bindings/bindings.types';
```

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+import type { Story } from '@muban/storybook';
+import {
+  bind,
+  defineComponent,
+  computed,
+  ref,
+  refCollection,
+  ComponentFactory,
+  ComponentRefItem,
+} from '@muban/muban';
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:11]:
```suggestion
import type { ComponentFactory, ComponentRefItem } from '@muban/muban';
import { bind, defineComponent, computed, ref, refCollection } from '@muban/muban';
```

File: test-storybook/src/components/bindings/css/Css.stories.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+export default {
+  title: 'bindings/css',
+};
+
+const cssTemplate = () => html` <div data-component="css">
+  <div data-ref="box" data-testid="box" class="box bg-primary">
+    <p>css: <span data-ref="info"></span></p>
+  </div>
+  <p>
+    <input data-ref="checkbox" data-testid="checkbox-box" type="checkbox" value="box" disabled />
```
Comment (ThaNarie) [inline test-storybook/src/components/bindings/css/Css.stories.ts:25]: The purpose of the `box` class (`class="box bg-primary"`) was to _not_ have it included in the bindings, to see if they would leave it alone completely. But with the `refs.checkboxes.getElements()` code in the stories, the `box` class is part of the binding.

So I would suggest to remove this `<p>`, and update the `play` tests that had `foobar-0` attached to `box`.

File: src/lib/bindings/dom/cssBinding.ts [hunk 8]
```diff
@@ -15,7 +15,9 @@ export function cssBinding(
     if (typeof classes === 'string') {
       // first remove the previously set classes, since our binding has changed
       const previousCssValue = (target as any)[previousCssBindingKey];
-      target.classList.remove(...previousCssValue.split(/\s/gi));
+      if (previousCssValue) {
+        target.classList.remove(...previousCssValue.split(/\s/gi));
+      }
 
       // then store and set our new classes
```
Comment (ThaNarie) [inline src/lib/bindings/dom/cssBinding.ts:22]: I observed an error in the console on the `target.classList.add(...classes.split(/\s/gi))` lines, when classes is an empty string (after removing the `box` input).

Doing `target.classList.add('')` throws a `DOMException`. I think that lines needs an `if`-statement as well :)

**Review by ThaNarie — CHANGES_REQUESTED** [review]
Nice progress, just a few more remarks, but we're close!

One general note (that you don't have to update here) but please do keep it in mind or any other Stories or Unit tests that you will be writing:

* Keep the Stories/Tests as simple as possible
* The current setup with the dynamic nature of `createCssStory`, `createCssComponent` and things like the dynamic fetching of the `classList` based on the `defaultValue`, could make it harder to follow what's going on
* Besides testing functionality, Stories/Tests are also meant as a "showcase / example" of how to use a certain feature. The simpler they look, the easier to follow.
* And lastly, the more complex your Story/Test, the harder it is to identify if – when they fail – the issue lies in the library code, or you Story/Test itself :D

It's a fine line to walk – reusing code can be fine, as long as it's easy to follow what's going on :)

File: test-storybook/src/components/bindings/css/Css.stories.ts (hunk 4 above)
Comment (jspolancor) [reply test-storybook/src/components/bindings/css/Css.stories.ts:52]: We tried to configure this but is quite tricky, the jest-dom import expects a globally defined **_expect_** variable to extend
image
But in storybook we're supossed to use the imported **_expect_** from the storybook/jest package
```javascript
import { expect } from '@storybook/jest';
```
There must be a way to make it work, but I think this dependency shouldn't be a blocker for this PR, we can use the expect from @storybook/jest and try to make it work in a specific issue for jest-dom installation

Comment (ThaNarie) [reply test-storybook/src/components/bindings/css/Css.stories.ts:52]: I got this to work by doing:

In `test-storybook/.storybook/preview.js`, add this at the top:
```ts
import { expect } from '@storybook/jest';
import * as matchers from '@testing-library/jest-dom/matchers';
expect.extend(matchers);
```

And then in your story file, you only have to import:
```ts
import { expect } from '@storybook/jest';
```

**Review by jspolancor — APPROVED** [review]

**Pull request merged**